11965 fixes that null value with unspecified nan_format crashes table rendering

harmbuisman opened the pull request:
- [X] issues: fixes #11965
- [X] tests added / passed: added unit test for issue in question
- [X] release document entry (if new feature or API change) - no change in behaviour

Note that this PR fixes the error when passing in a null value to the ScientificFormatter. The output is now the exact same as when you use a NumberFormatter:
![image](https://user-images.githubusercontent.com/7702207/165106115-7b3ecfd7-5470-4de9-b988-f31fd4fb613c.png)

bryevdv commented:
Couldn't this also be handled by making there be a default nan formatter? I guess I would prefer that to making a special case.

harmbuisman commented:
> Couldn't this also be handled by making there be a default nan formatter? (and possibly also making the formatter value non-nullable) I guess I would prefer that to making a special case.
> 
> Edit: and to avoid surprises and "why is this field empty" support questions I also think I would prefer the default nan format to visible, e.g. "NaN" then if people want a blank cell they can choose that explicitly.

Hi, I missed your edit before making the changes. I saw code duplication so moved the logic to StringFormatter. Since it is used by NumberFormatter, ScientificFormatter and DateFormatter "NaN" may not be a good default. For DateFormatter NaT may be better. Also, I noticed Numbro uses NaN as the default, so to preserve existing behaviour I NaN is indeed better. Maybe I should revert the change of moving the setting to StringFormatter? Or is it possible to override the default in the case of DateFormatter.

It is a bit weird that there is a row with NaNs even though the source is empty:
![image](https://user-images.githubusercontent.com/7702207/165166333-21e747f0-8cd3-46e0-aa78-f6c9ae0f7346.png)
I guess better is to just show the header? Need to find out where the table is drawn.

bryevdv commented:
> It is a bit weird that there is a row with NaNs even though the source is empty

This just seems like a bug. Does something this happen without the changes in this PR?

bryevdv commented:
@bokeh/core any strong feelings about "" vs "NaN" for the default ?

harmbuisman commented:
> > It is a bit weird that there is a row with NaNs even though the source is empty
> 
> This just seems like a bug. Does something this happen without the changes in this PR?

This is not caused by the changes in the PR. So likely better to create a separate issue for it.

bryevdv commented:
@philippjfr does Panel/HV have an preferences for default formatting of NaN in tables?

jbednar commented:
My vote is for "-", given that NaN and null are both tied to specific languages rather than being user-level concepts.

philippjfr commented:
Yes in all my client applications I set the NaN formatter to "-".

bryevdv commented:
@harmbuisman Does "-" work for you? If so can you make that change then we can merge this

harmbuisman commented:
> @harmbuisman Does "-" work for you? If so can you make that change then we can merge this

Yes, that works. That is appropriate for all, including DateFormatter. I updated all unit tests that used "-" as nan_format input value and changed it to "--" so that it tests something different than the default value.

harmbuisman commented:
@bryevdv ,

I checked the failed tests. I cannot relate the error message to the changes in this PR:
selenium.common.exceptions.TimeoutException: Message: Connection refused (os error 61)

Maybe a glitch?

bryevdv commented:
Thanks @harmbuisman !
